<div style="text-align:center; color:red;">
    <h2>Retail Sales Case Study</h2>
</div>

<div style="text-align: justify; background-color:#e0f2f1; padding:10px">We are given a dataset from a Retail gaint. The dataset contains information on : <br> <p>1. Demographics of the customers, <b><i>demographics.txt</i></b> <br> 2. Purchase Behaviour, <b><i>behaviour.json</i></b> <br>
3. Response to various marketing campaigns run, <b><i>campaign.json</i></b></p> </div>

<div style="text-align: justify; background-color:#dcedc8; padding:10px"><h3>Business Problem</h3> <br>The retailer wants to understand what kind of customers respond to different campaigns. To arrive at a reasonable answer to the above question, you've been tasked to analyze this dataset. Below are some pointed business questions you are required to answer. </div>

<div style="text-align: justify; background-color:#dcedc8; padding:10px">
    <h3>Data Quality and Check (Task 1)</h3>
<br>1.
Create a consolidated view of data by joining the data present in three files.
<br>2.
Are there any variables where you will need to clean the raw data, what kind of cleaning will be needed?
<br>3.
Create a data quality report after doing the necessary cleaning and joining of the files by:
<br> &nbsp;&nbsp;&nbsp;•
Doing univariates for continuous variables (compute: percentage of missing values, percentage of terms which are zero, mean, 25th, 50th, 75th, 90th and 95th percentile, min and max)
<br> &nbsp;&nbsp;&nbsp;•
Doing univariates for categorical variables (compute:percentage of missing values, number of unique values)
<br>4.
Are there any extreme values of variables representing income, amount of money spent on various categories, recency of purchase?</div>

<div style="text-align: justify; background-color:#fff9c4; padding:10px"><b>1.</b> To create a consolidated view of data by joining the data present in three files (two JSON files and one text file), can follow these general steps:<br>
<b>Read the Files:</b> Begin by reading the data from each of the three files into Python script.<br>
<b>Parse JSON Files:</b> Parse the data from the JSON files into Python dictionaries or objects.<br>
<b>Process Text File:</b> Read and process the data from the text file according to its format.<br>
<b>Combine Data:</b> Once you have the data from all three sources, combine them based on a common identifier.<br>
<b>Consolidate Data:</b> Consolidate the combined data into a single data structure, such as a list of dictionaries or a pandas DataFrame.<br>
<b>Output the Consolidated Data:</b> Depending on requirement, we can output the consolidated data in a CSV format.</div>

In [1]:
# Importing Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

<h2 style="color:blue"><i>For behaviour.json</h2>

In [2]:
# importing JSON module
import json

# Read data of BEHAVIOUR.JSON file (file should be in same directory)
with open("behaviour.json",'r') as f1:
   data_1 = json.load(f1)
#print(data_1) # usually data looks in json file
#data_1 # in more readable manner

<div style="text-align: justify; background-color:#fff9c4; padding:10px">After seeing data in <b><i>`behaviour.json`</i></b> file we realized that the data stored as list of dictionaries, where each dictionary represents a customer's data. Each dictionary contains various key-value pairs representing different attributes of the customer, such as their ID, recency, purchase amounts for different product categories (e.g., wines, fruits, meat products), number of purchases through different channels (e.g., web, catalog, store), and number of web visits per month.</div>

In [3]:
# Create an empty list to store data of behavior.json
rows = []

# Loop through each dictionary in the list
for items in data_1:
    # Extract the key (ID) and the corresponding dictionary
    id_ = list(items.keys())[0]
    values = items[id_]

    # Create a dictionary with Id as a key and value
    row = {'ID':id_}
    row.update(values)
    rows.append(row)

# Creating DataFrame
df_behaviour = pd.DataFrame(rows)

In [4]:
# Display DataFrame
print(df_behaviour.head())

         ID  Recency  MntWines  MntFruits  MntMeatProducts  MntFishProducts  \
0   ID_1826        0       189        104              379              111   
1      ID_1        0       464          5               64                7   
2  ID_10476        0       134         11               59               15   
3   ID_1386        0        10          0                1                0   
4   ID_5371        0         6         16               24               11   

   MntSweetProducts  MntGoldProds  NumDealsPurchases  NumWebPurchases  \
0               189           218                  1                4   
1                 0            37                  1                7   
2                 2            30                  1                3   
3                 0             0                  1                1   
4                 0            34                  2                3   

   NumCatalogPurchases  NumStorePurchases  NumWebVisitsMonth  
0                    4 

In [5]:
# Remove the 'ID_' prefix from the key 'Id' and store the result
df_behaviour["ID"] = df_behaviour['ID'].str.replace('ID_','')
print(df_behaviour.head()) 

      ID  Recency  MntWines  MntFruits  MntMeatProducts  MntFishProducts  \
0   1826        0       189        104              379              111   
1      1        0       464          5               64                7   
2  10476        0       134         11               59               15   
3   1386        0        10          0                1                0   
4   5371        0         6         16               24               11   

   MntSweetProducts  MntGoldProds  NumDealsPurchases  NumWebPurchases  \
0               189           218                  1                4   
1                 0            37                  1                7   
2                 2            30                  1                3   
3                 0             0                  1                1   
4                 0            34                  2                3   

   NumCatalogPurchases  NumStorePurchases  NumWebVisitsMonth  
0                    4                  6

In [6]:
# saving DataFrame to a CSV file
df1 = df_behaviour.to_csv('behaviour.csv', index=False)

<h2 style="color:blue"><i>For campaign.json</h2>

In [3]:
# Read data of CAMPAIGN.JSON file (file should be in same directory as file)
with open("campaign.json",'r') as f2:
   data_2 = json.load(f2)
#print(data_2) # usually data looks in json file
#data_2 # in more readable manner

<div style="text-align: justify; background-color:#fff9c4; padding:10px">After seeing data in <b><i>`campaign.json`</i></b> file we realized that the data stored as list of dictionaries, where each dictionary represents a customer's data.<br> Each dictionary contains various key-value pairs representing different Attributes related to campaign acceptance, such as 'AcceptedCmp1', 'AcceptedCmp2', 'AcceptedCmp3', 'AcceptedCmp4', and 'AcceptedCmp5'.<br>
Attributes related to other responses, such as 'Response' and 'Complain'.</div>

In [8]:
# Initailize an empty list to store data of behavior.json
rows = []

# Iterate over each dictionary in the list
for items in data_2:
    # Extract the Id and values from the dictionary
    id_ = list(items.keys())[0]
    values = items[id_]

    # Create a dictionary with Id as a key and value
    row = {'ID':id_}
    row.update(values)
    rows.append(row)

# Creating DataFrame
df_campaign = pd.DataFrame(rows)

In [9]:
# Display DataFrame
print(df_campaign.head())

         ID  AcceptedCmp1  AcceptedCmp2  AcceptedCmp3  AcceptedCmp4  \
0   ID_1826             0             0             0             0   
1      ID_1             0             1             0             0   
2  ID_10476             0             0             0             0   
3   ID_1386             0             0             0             0   
4   ID_5371             0             0             1             0   

   AcceptedCmp5  Response  Complain  
0             0         1         0  
1             0         1         0  
2             0         0         0  
3             0         0         0  
4             0         1         0  


In [10]:
# Remove the 'ID_' prefix from the key 'Id' and store the result
df_campaign["ID"] = df_campaign['ID'].str.replace('ID_','')
print(df_campaign.head()) 

      ID  AcceptedCmp1  AcceptedCmp2  AcceptedCmp3  AcceptedCmp4  \
0   1826             0             0             0             0   
1      1             0             1             0             0   
2  10476             0             0             0             0   
3   1386             0             0             0             0   
4   5371             0             0             1             0   

   AcceptedCmp5  Response  Complain  
0             0         1         0  
1             0         1         0  
2             0         0         0  
3             0         0         0  
4             0         1         0  


In [11]:
# saving DataFrame to a CSV file
df2 = df_campaign.to_csv('campaign.csv', index=False)

<h2 style="color:blue"><i>For demographics.txt</h2> 

In [12]:
# Read the text file into a DataFrame
df_demographics = pd.read_csv('demographics.txt', delimiter='\t')

# Display the DataFrame
print(df_demographics.head())

      ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
0   1826        1970  Graduation       Divorced  $84,835.00         0   
1      1        1961  Graduation         Single  $57,091.00         0   
2  10476        1958  Graduation        Married  $67,267.00         0   
3   1386        1967  Graduation       Together  $32,474.00         1   
4   5371        1989  Graduation         Single  $21,474.00         1   

   Teenhome Dt_Customer Country  
0         0     6/16/14      SP  
1         0     6/15/14      CA  
2         1     5/13/14      US  
3         1     5/11/14     AUS  
4         0      4/8/14      SP  


<div style="text-align: justify; background-color:#fff9c4; padding:10px">After seeing data in <b><i>`demographics.txt`</i></b> file we realized that the data appears to be structured tabular data with various columns. Each row represents information about a customer, and each column represents a specific attribute of the customer, such as ID, birth year, education, marital status, income, number of kids at home, number of teenagers at home, date of customer registration, and country.</div>

In [13]:
# Saving DataFrame to a CSV file
df3 = df_demographics.to_csv('demographics.csv', index=False)

<h2 style="color:blue"><i>Merging data</h2> 

In [14]:
# Check the data type of the 'ID' column of behaviour file
id_dtype1 = df_behaviour['ID'].dtype
print("Data type of 'ID' column in behaviour file:", id_dtype1)

# Check the data type of the 'ID' column of campaign file
id_dtype2 = df_campaign['ID'].dtype
print("Data type of 'ID' column in campaign file:", id_dtype2)

# Check the data type of the 'ID' column of demographics file
id_dtype3 = df_demographics['ID'].dtype
print("Data type of 'ID' column in demographics file:", id_dtype3)

Data type of 'ID' column in behaviour file: object
Data type of 'ID' column in campaign file: object
Data type of 'ID' column in demographics file: int64


<div style="text-align: justify; background-color:#fff9c4; padding:10px">Since the 'ID' column has different data types in each file, we need to ensure consistency before merging. We'll convert the 'ID' column in the behaviour and campaign files to the same data type as the demographics file (int64) before merging.

In [15]:
# Convert 'ID' column to int64 of the behaviour file
df_behaviour['ID'] = df_behaviour['ID'].astype(int)

# Convert 'ID' column to int64 of the campaign file 
df_campaign['ID'] = df_campaign['ID'].astype(int)

# Merge the dataframes
df_merged = pd.merge(df_demographics, df_behaviour, on='ID', how='inner')
df_merged = pd.merge(df_merged, df_campaign, on='ID', how='inner')

# Save the merged dataframe to a CSV file
df_merged.to_csv('merged_data.csv', index=False)

# Display merged data
print(df_merged.head())

      ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
0   1826        1970  Graduation       Divorced  $84,835.00         0   
1      1        1961  Graduation         Single  $57,091.00         0   
2  10476        1958  Graduation        Married  $67,267.00         0   
3   1386        1967  Graduation       Together  $32,474.00         1   
4   5371        1989  Graduation         Single  $21,474.00         1   

   Teenhome Dt_Customer Country  Recency  ...  NumCatalogPurchases  \
0         0     6/16/14      SP        0  ...                    4   
1         0     6/15/14      CA        0  ...                    3   
2         1     5/13/14      US        0  ...                    2   
3         1     5/11/14     AUS        0  ...                    0   
4         0      4/8/14      SP        0  ...                    1   

   NumStorePurchases  NumWebVisitsMonth  AcceptedCmp1  AcceptedCmp2  \
0                  6                  1             0             0  

<div style="text-align: justify; background-color:#e6e6fa; padding:10px"><h3>2. To determine if there are any variables in the raw data that require cleaning, we need to assess the data for potential issues such as missing values, inconsistencies, outliers, or formatting errors.

In [16]:
# Read data from merged_data.csv file
data = pd.read_csv('merged_data.csv')

# Trim column names
data.columns = data.columns.str.strip()

# Display first 5 rows of the data
data.head()

ID  Year_Birth   Education Marital_Status       Income  Kidhome  \
0   1826        1970  Graduation       Divorced  $84,835.00         0   
1      1        1961  Graduation         Single  $57,091.00         0   
2  10476        1958  Graduation        Married  $67,267.00         0   
3   1386        1967  Graduation       Together  $32,474.00         1   
4   5371        1989  Graduation         Single  $21,474.00         1   

   Teenhome Dt_Customer Country  Recency  ...  NumCatalogPurchases  \
0         0     6/16/14      SP        0  ...                    4   
1         0     6/15/14      CA        0  ...                    3   
2         1     5/13/14      US        0  ...                    2   
3         1     5/11/14     AUS        0  ...                    0   
4         0      4/8/14      SP        0  ...                    1   

   NumStorePurchases  NumWebVisitsMonth  AcceptedCmp1  AcceptedCmp2  \
0                  6                  1             0             0   
1                  7                  5             0             1   
2                  5                  2             0             0   
3                  2                  7             0             0   
4                  2                  7             0             0   

   AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  Response  Complain  
0             0             0             0         1         0  
1             0             0             0         1         0  
2             0             0             0         0         0  
3             0             0             0         0         0  
4             1             0             0         1         0  

[5 rows x 28 columns]

In [17]:
# Data Types and Value Counts
dtype_counts = data.dtypes.value_counts()
print("Datatypes in the dataset and their respective value counts:")
print(dtype_counts)
print("-" * 60)

# Detailed breakdown of data types with column names
print("Data Type: int64")
print(data.select_dtypes(include='int64').info())
print("-" * 60)
print("Data Type: object")
print(data.select_dtypes(include='object').info())

Datatypes in the dataset and their respective value counts:
int64     23
object     5
dtype: int64
------------------------------------------------------------
Data Type: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   ID                   2240 non-null   int64
 1   Year_Birth           2240 non-null   int64
 2   Kidhome              2240 non-null   int64
 3   Teenhome             2240 non-null   int64
 4   Recency              2240 non-null   int64
 5   MntWines             2240 non-null   int64
 6   MntFruits            2240 non-null   int64
 7   MntMeatProducts      2240 non-null   int64
 8   MntFishProducts      2240 non-null   int64
 9   MntSweetProducts     2240 non-null   int64
 10  MntGoldProds         2240 non-null   int64
 11  NumDealsPurchases    2240 non-null   int64
 12  NumWebPurchases      2240 non-null   in

<div style="text-align: justify; background-color:#e6e6fa; padding:10px">The above output presents a summary of the dataset's data types and their respective value counts, followed by a detailed breakdown of each data type category (int64 and object) with column names, non-null counts, and data types.

<b>1. Datatypes in the dataset and their respective value counts:</b><br>
    &nbsp;&nbsp;&nbsp;It indicates that there are <i>23 columns</i> with data type <i>int64</i> and <i>5 columns</i> with data type <i>object</i>.<br>
&nbsp;&nbsp;&nbsp;The value counts for each data type are presented in a Series format.<br>
<b>2. Detailed breakdown of data types with column names:</b><br>
    <i>For columns of data type int64:</i><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; It displays a DataFrame summary with 23 columns.<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Each column's name, non-null count, and data type (Dtype) are provided.<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; The summary is followed by memory usage information.<br>
    <i>For columns of data type object:</i><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; It displays another DataFrame summary with 5 columns.<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Similar to int64, it presents column names, non-null counts, and data types (Dtype).<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Memory usage information is also provided.

In [18]:
# Check for missing values
missing_values = data.isnull().sum()

# Calculate the percentage of missing values
missing_percentage = (missing_values / len(data)) * 100

# Create a DataFrame to display missing values and their percentage
missing_info = pd.DataFrame({'Missing Values': missing_values, 'Percentage': missing_percentage})

# Print the missing information
print("Missing Values and Percentage:")
print(missing_info)

Missing Values and Percentage:
                     Missing Values  Percentage
ID                                0    0.000000
Year_Birth                        0    0.000000
Education                         0    0.000000
Marital_Status                    0    0.000000
Income                           24    1.071429
Kidhome                           0    0.000000
Teenhome                          0    0.000000
Dt_Customer                       0    0.000000
Country                           0    0.000000
Recency                           0    0.000000
MntWines                          0    0.000000
MntFruits                         0    0.000000
MntMeatProducts                   0    0.000000
MntFishProducts                   0    0.000000
MntSweetProducts                  0    0.000000
MntGoldProds                      0    0.000000
NumDealsPurchases                 0    0.000000
NumWebPurchases                   0    0.000000
NumCatalogPurchases               0    0.000000
NumStoreP

<div style="text-align: justify; background-color:#e6e6fa; padding:10px">It appears that the values in <b>'Income'</b> are formatted as currency with a dollar sign ('$') and commas (',') as thousand separators.<br>
To convert these values to numeric format, we need to remove the dollar sign and commas from each value before converting them to numeric.

In [19]:
# Remove dollar sign and commas, convert to float, and then to int
data['Income'] = data['Income'].str.replace('$', '').str.replace(',', '').astype(float).astype('Int64')

# Print the DataFrame to verify the update
data['Income']

C:\Users\Dell\AppData\Local\Temp\ipykernel_17372\4244439147.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data['Income'] = data['Income'].str.replace('$', '').str.replace(',', '').astype(float).astype('Int64')


0       84835
1       57091
2       67267
3       32474
4       21474
        ...  
2235    66476
2236    31056
2237    46310
2238    65819
2239    94871
Name: Income, Length: 2240, dtype: Int64

In [20]:
# Calculate the mean of the 'Income' column
income_mean = int(data['Income'].mean())

# Print the mean of the 'Income' column
print("Mean of Income:", income_mean)

# Fill missing values in 'Income' column with mean
data['Income'].fillna(income_mean, inplace=True)

# Print the DataFrame to verify the update
print(data.head())

Mean of Income: 52247
      ID  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       Divorced   84835        0         0   
1      1        1961  Graduation         Single   57091        0         0   
2  10476        1958  Graduation        Married   67267        0         1   
3   1386        1967  Graduation       Together   32474        1         1   
4   5371        1989  Graduation         Single   21474        1         0   

  Dt_Customer Country  Recency  ...  NumCatalogPurchases  NumStorePurchases  \
0     6/16/14      SP        0  ...                    4                  6   
1     6/15/14      CA        0  ...                    3                  7   
2     5/13/14      US        0  ...                    2                  5   
3     5/11/14     AUS        0  ...                    0                  2   
4      4/8/14      SP        0  ...                    1                  2   

   NumWebVisitsMonth  AcceptedCmp1

In [21]:
# After replace missing values in 'Income' column confirm again for missing values
missing_val = data.isnull().sum()
print("Missing Values:")
print(missing_val)

Missing Values:
ID                     0
Year_Birth             0
Education              0
Marital_Status         0
Income                 0
Kidhome                0
Teenhome               0
Dt_Customer            0
Country                0
Recency                0
MntWines               0
MntFruits              0
MntMeatProducts        0
MntFishProducts        0
MntSweetProducts       0
MntGoldProds           0
NumDealsPurchases      0
NumWebPurchases        0
NumCatalogPurchases    0
NumStorePurchases      0
NumWebVisitsMonth      0
AcceptedCmp1           0
AcceptedCmp2           0
AcceptedCmp3           0
AcceptedCmp4           0
AcceptedCmp5           0
Response               0
Complain               0
dtype: int64


<div style="text-align: justify; background-color:#e6e6fa; padding:10px">We replaced 24 missing values in income with the mean of the income. Hence, there is no missing values as of now.

In [22]:
# Data Consistency
# Check for consistency in categorical variables
categorical_variables = ['Education', 'Marital_Status', 'Country']
for col in categorical_variables:
    print(f"Unique values in {col}:")
    print(data[col].unique())

Unique values in Education:
['Graduation' 'PhD' '2n Cycle' 'Master' 'Basic']
Unique values in Marital_Status:
['Divorced' 'Single' 'Married' 'Together' 'Widow' 'YOLO' 'Alone' 'Absurd']
Unique values in Country:
['SP' 'CA' 'US' 'AUS' 'GER' 'IND' 'SA' 'ME']


In [23]:
# Standardize 'Marital_Status' column
data['Marital_Status'] = data['Marital_Status'].str.upper()

# Check for duplicates in 'ID' column
duplicate_ids = data[data.duplicated(subset=['ID'])]
print("Duplicate IDs:")
print(duplicate_ids)

# Display summary after cleaning
print("Summary after cleaning:")
print(data.info())

Duplicate IDs:
Empty DataFrame
Columns: [ID, Year_Birth, Education, Marital_Status, Income, Kidhome, Teenhome, Dt_Customer, Country, Recency, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, NumDealsPurchases, NumWebPurchases, NumCatalogPurchases, NumStorePurchases, NumWebVisitsMonth, AcceptedCmp1, AcceptedCmp2, AcceptedCmp3, AcceptedCmp4, AcceptedCmp5, Response, Complain]
Index: []

[0 rows x 28 columns]
Summary after cleaning:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   ID                   2240 non-null   int64 
 1   Year_Birth           2240 non-null   int64 
 2   Education            2240 non-null   object
 3   Marital_Status       2240 non-null   object
 4   Income               2240 non-null   Int64 
 5   Kidhome              2240 non-null   int64 
 6   Teenhome             2240

In [24]:
data.head()

ID  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
0   1826        1970  Graduation       DIVORCED   84835        0         0   
1      1        1961  Graduation         SINGLE   57091        0         0   
2  10476        1958  Graduation        MARRIED   67267        0         1   
3   1386        1967  Graduation       TOGETHER   32474        1         1   
4   5371        1989  Graduation         SINGLE   21474        1         0   

  Dt_Customer Country  Recency  ...  NumCatalogPurchases  NumStorePurchases  \
0     6/16/14      SP        0  ...                    4                  6   
1     6/15/14      CA        0  ...                    3                  7   
2     5/13/14      US        0  ...                    2                  5   
3     5/11/14     AUS        0  ...                    0                  2   
4      4/8/14      SP        0  ...                    1                  2   

   NumWebVisitsMonth  AcceptedCmp1  AcceptedCmp2  AcceptedCmp3  AcceptedCmp4  \
0                  1             0             0             0             0   
1                  5             0             1             0             0   
2                  2             0             0             0             0   
3                  7             0             0             0             0   
4                  7             0             0             1             0   

   AcceptedCmp5  Response  Complain  
0             0         1         0  
1             0         1         0  
2             0         0         0  
3             0         0         0  
4             0         1         0  

[5 rows x 28 columns]

In [25]:
# Standardize the 'Dt_Customer' column to the 'mm/dd/yy' format
data['Dt_Customer'] = pd.to_datetime(data['Dt_Customer'], errors='coerce')

# Check for any dates that could not be parsed
invalid_dates = data[data['Dt_Customer'].isnull()]['Dt_Customer']
print("Invalid dates:")
print(invalid_dates)

# Drop rows with invalid dates
data.dropna(subset=['Dt_Customer'], inplace=True)

# Display summary after cleaning
print("Summary after cleaning:")
print(data['Dt_Customer'].info())

Invalid dates:
Series([], Name: Dt_Customer, dtype: datetime64[ns])
Summary after cleaning:
<class 'pandas.core.series.Series'>
RangeIndex: 2240 entries, 0 to 2239
Series name: Dt_Customer
Non-Null Count  Dtype         
--------------  -----         
2240 non-null   datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 17.6 KB
None


<div style="text-align: justify; background-color:#e6e6fa; padding:10px">This above code aims to standardize the date format in the 'Dt_Customer' column and handle any invalid dates that may be present in the dataset.

<div style="text-align: justify; background-color:#fff9c4; padding:10px">
<h3>3. To create a data quality report with univariate analysis for both continuous and categorical variables: </h3>

***Continuous Variables:*** 
+ Calculate the percentage of missing values.
+ Compute the percentage of terms which are zero.
+ Calculate descriptive statistics (mean, percentiles, min, max).
+ Generate a summary report.

***Categorical Variables:***
+ Calculate the percentage of missing values.
+ Determine the number of unique values.
+ Generate a summary report.

In [31]:
# Select the Continuous data
cont = data.select_dtypes(include=np.number)
print("Continuous Data :")
print(cont.columns)

#Select the Categorical data
catg = data.select_dtypes(include=object)
print("\n""Categorical Data")
print(catg.columns)

Continuous Data :
Index(['ID', 'Year_Birth', 'Income', 'Kidhome', 'Teenhome', 'Recency',
       'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts',
       'MntSweetProducts', 'MntGoldProds', 'NumDealsPurchases',
       'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases',
       'NumWebVisitsMonth', 'AcceptedCmp1', 'AcceptedCmp2', 'AcceptedCmp3',
       'AcceptedCmp4', 'AcceptedCmp5', 'Response', 'Complain'],
      dtype='object')

Categorical Data
Index(['Education', 'Marital_Status', 'Country'], dtype='object')


In [43]:
# Continuous Variables Analysis
continuous_vars = ['Income', 'Kidhome', 'Teenhome', 'Recency', 'MntWines', 'MntFruits', 
                   'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts', 'MntGoldProds', 
                   'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases', 
                   'NumWebVisitsMonth', 'AcceptedCmp1', 'AcceptedCmp2', 'AcceptedCmp3', 'AcceptedCmp4', 
                   'AcceptedCmp5', 'Response', 'Complain']

## Calculate statistics for continuous variables
continuous_stats = data[continuous_vars].describe(percentiles=[0.25, 0.5, 0.75, 0.90, 0.95])

# Display the analysis
print("Continuous Variables Analysis:""\n")
print(continuous_stats)

Continuous Variables Analysis:

             Income      Kidhome     Teenhome      Recency     MntWines  \
count        2240.0  2240.000000  2240.000000  2240.000000  2240.000000   
mean   52247.248661     0.444196     0.506250    49.109375   303.935714   
std    25037.797168     0.538398     0.544538    28.962453   336.597393   
min          1730.0     0.000000     0.000000     0.000000     0.000000   
25%        35538.75     0.000000     0.000000    24.000000    23.750000   
50%         51741.5     0.000000     0.000000    49.000000   173.500000   
75%        68289.75     1.000000     1.000000    74.000000   504.250000   
90%         79800.3     1.000000     1.000000    89.000000   822.100000   
95%         83927.0     1.000000     1.000000    94.000000  1000.000000   
max        666666.0     2.000000     2.000000    99.000000  1493.000000   

         MntFruits  MntMeatProducts  MntFishProducts  MntSweetProducts  \
count  2240.000000      2240.000000      2240.000000       2240.0000

<div style="text-align: justify; background-color:#fff9c4; padding:10px">The above table provides descriptive statistics for various continuous variables in your dataset. Here's what each statistic represents:<br>
    
+ count: This represents the number of valid (non-missing) observations for each variable. It indicates the size of the dataset for that particular variable.
+ mean: The mean (average) of a variable is calculated by summing all the values of the variable and dividing by the total number of observations. It gives you a measure of the central tendency of the data.
+ std: The standard deviation is a measure of the dispersion or spread of the values around the mean. A higher standard deviation indicates greater variability in the data, while a lower standard deviation indicates less variability.
+ min: This is the smallest value observed in the dataset for each variable. It represents the lower bound of the range of values.
+ 25% (1st Quartile): Also known as the first quartile or lower quartile, this value indicates the point below which 25% of the observations fall. It's the value that separates the lowest 25% of the data from the rest.
+ 50% (2nd Quartile/Median): The median is the middle value of the dataset when the observations are arranged in ascending order. It represents the value below which 50% of the observations fall.
+ 75% (3rd Quartile): The third quartile or upper quartile indicates the point below which 75% of the observations fall. It's the value that separates the lowest 75% of the data from the highest 25%.
+ 90%: This value indicates the point below which 90% of the observations fall. It provides insight into the distribution of the data towards the lower end.
+ 95%: Similar to the 90th percentile, this value indicates the point below which 95% of the observations fall, providing further insight into the distribution of the data towards the lower end.
+ max: This is the largest value observed in the dataset for each variable. It represents the upper bound of the range of values.

<div style="text-align: justify; background-color:#fff9c4; padding:10px">The above table provides descriptive statistics for various continuous variables in your dataset. Here's what each statistic represents:<br>
    
+ **count:** The number of non-missing values for each variable.
+ **mean:** The average value of each variable.
+ **std:** The standard deviation, which measures the dispersion or spread of the values.
+ **min:** The minimum value observed for each variable.
+ **25%:** The 25th percentile (first quartile), indicating the value below which 25% of the data fall.
+ **50%:** The median (second quartile), representing the middle value when the data is sorted in ascending order.
+ **75%:** The 75th percentile (third quartile), indicating the value below which 75% of the data fall.
+ **90%:** The 90th percentile, indicating the value below which 90% of the data fall.
+ **95%:** The 95th percentile, indicating the value below which 95% of the data fall.
+ **max:** The maximum value observed for each variable.

*For example,* in the "Income" variable, the mean income is approximately **52,247.25**, with a standard deviation of approximately **25,037.80**. The minimum income observed is **1,730**, and the maximum income observed is **666,666**.

In [45]:
# Calculate percentage of missing values
missing_values_cat = catg.isnull().mean() * 100

# Determine number of unique values
unique_values = catg.nunique()

# Combine the results
categorical_analysis = pd.concat([missing_values_cat, unique_values], axis=1)
categorical_analysis.columns = ['% Missing Values', 'Unique Values']

# Display the analysis
print("Categorical Variables Analysis:\n")
print(categorical_analysis)

Categorical Variables Analysis:

                % Missing Values  Unique Values
Education                    0.0              5
Marital_Status               0.0              8
Country                      0.0              8


<div style="text-align: justify; background-color:#fff9c4; padding:10px">Above analysis provides insights into the categorical variables in your dataset:

1. **% Missing Values**: This indicates the percentage of missing values (if any) in each categorical variable. A missing value implies that the information for that variable is not available for certain observations in your dataset. In this case, there are no missing values for any of the categorical variables, as the percentage is 0.0 for all of them.

2. **Unique Values**: This represents the count of unique categories or levels within each categorical variable. It gives you an idea of the diversity or variability present in the data for each categorical variable. For example:
   - **Education**: There are 5 unique values or levels, indicating 5 different education categories among the observations.
   - **Marital_Status**: There are 8 unique values or levels, suggesting 8 different marital status categories.
   - **Country**: There are 8 unique values or levels, implying that the dataset includes data from 8 different countries.


<div style="text-align: justify; background-color:#e6e6fa; padding:10px"><h3>4. 
To identify extreme values or outliers in variables representing income, amount of money spent on various categories, and recency of purchase :</h3>

1. **Income**: Look for values that are significantly higher or lower than the majority of the data. You can calculate the interquartile range (IQR) and then identify values that are more than 1.5 times the IQR above the third quartile (Q3) or below the first quartile (Q1).

2. **Amount of money spent on various categories**: Follow a similar approach as with income. Calculate the IQR for each spending category and identify values that fall outside the 1.5 times IQR range above Q3 or below Q1.

3. **Recency of purchase**: Examine values that are significantly higher than the typical range of recency values. This could indicate customers who haven't made purchases for an unusually long time.


In [52]:
# Define the continuous variables of interest
cont_vars = ['Income', 'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts',
                   'MntSweetProducts', 'MntGoldProds', 'Recency']

# Calculate the interquartile range (IQR) for each continuous variable
Q1 = data[cont_vars].quantile(0.25)
Q3 = data[cont_vars].quantile(0.75)
IQR = Q3 - Q1

# Identify outliers for each continuous variable
outliers = (data[cont_vars] < (Q1 - 1.5 * IQR)) | (data[cont_vars] > (Q3 + 1.5 * IQR))

# Print the count of outliers for each conti_vars
for var in cont_vars:
    print(f"Number of outliers for {var}: {outliers[var].sum()}")

Number of outliers for Income: 8
Number of outliers for MntWines: 35
Number of outliers for MntFruits: 227
Number of outliers for MntMeatProducts: 175
Number of outliers for MntFishProducts: 223
Number of outliers for MntSweetProducts: 248
Number of outliers for MntGoldProds: 207
Number of outliers for Recency: 0


<div style="text-align: justify; background-color:#e6e6fa; padding:10px">This code is performing outlier detection for a set of continuous variables in a DataFrame. Here's a breakdown of each step:

1. **Define Continuous Variables**: First, a list named `cont_vars` is defined, containing the names of continuous variables of interest.

2. **Calculate Interquartile Range (IQR)**: The Interquartile Range (IQR) is calculated for each continuous variable using the `quantile` method with a parameter of 0.25 (Q1) and 0.75 (Q3) to get the first and third quartiles, respectively. Then, the IQR is computed as the difference between Q3 and Q1.

3. **Identify Outliers**: Outliers for each continuous variable are identified using the IQR method. For each variable, a Boolean DataFrame is created where `True` indicates the presence of an outlier, calculated based on whether the value falls below `(Q1 - 1.5 * IQR)` or above `(Q3 + 1.5 * IQR)`.

4. **Print Outlier Counts**: Finally, a loop iterates over each continuous variable in `cont_vars`. For each variable, it prints the count of outliers found by summing up the `True` values in the Boolean DataFrame created in the previous step for that variable.

This process provides insight into whether there are any extreme values (outliers) for each selected continuous variable in the dataset.